File: algo1/1/inv.py
import sys
import os
import time

inv = 0;

def merge(left, right):
    global inv
    result = []
    i ,j = 0, 0
    bi = len(left)
    bj = len(right)

    for k in range(bi+bj):
        if j >= bj or ( i < bi and left[i] <= right[j] ):
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
            inv += (bi - i)

    return result

def mergesort(list):
    l = len(list)
    if l < 2:
        return list
    else:
        middle = l / 2
        left = mergesort(list[:middle])
        right = mergesort(list[middle:])
        return merge(left, right)

t0 = time.clock()

lines = [int(line) for line in open('IntegerArray.txt')]
mergesort(lines)
print 'inversions: ' + str(inv)

print 'time: ' + str(time.clock()-t0)

File: algo1/1/inv.php
<?php


function countsort( $a, $n ) {
    if( $n == 1 ) return $a;

    $nb = $n >> 1; $nc = $n - $nb;

    $b = array_slice( $a, 0, $nb );
    $c = array_slice( $a, -$nc );

    $b = countsort( $b, $nb );
    $c = countsort( $c, $nc );

    return countmerge( $b, $c, $nb, $nc );
}

function countmerge( $b, $c, $nb, $nc ) {
    $i = $j = 0; $max = $nb + $nc;
    $d = array();

    for( $k = 0; $k < $max; $k++ ) {
        if( $j >= $nc or ( $i < $nb and $b[$i] <= $c[$j] ) ) {
            $d[] = $b[$i];
            $i++;
        } else {
            $d[] = $c[$j];
            $j++;
            global $inv;
            $inv += ( $nb - $i );
        }
    }

    return $d;
}

$time_start = microtime(true);

$a = file('IntegerArray.txt');
function toInt($x) {
    return (int)$x;
}
$a = array_map( 'toInt', $a );
$n = count( $a );
$inv = 0;
$d = countsort( $a, $n );

$time_end = microtime(true);
$time = $time_end - $time_start;

echo 'inversions: ' . $inv . "\n";
echo 'time: ' . $time . "\n";

?>